### ISSUE TITLE
JUnit Platform + Jmockit won't work

### ISSUE THREAD

**Issue by MichalMichalak** [gradle/gradle#4618]
When using new JUnit Platform support and Jmockit, mocks are not working properly.
* JMockit 1.38
* JUnit5 Jupiter 5.1.0
* Gradle 4.6

### Expected Behavior
Mocked method with recorded expectations should return defined values.

### Current Behavior
They are returning `null`.

### Context
I've opened JMockit issue for this, but the author suggested problem is in JUnit Platform implementation. Specifically in the class: https://github.com/gradle/gradle/blob/master/subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestClassProcessor.java
Here's the link: https://github.com/jmockit/jmockit1/issues/508#issuecomment-370910563

#### Build script
```
plugins {
    id 'java'
}

tasks.withType(JavaCompile) {
    sourceCompatibility = JavaVersion.VERSION_1_8
    targetCompatibility = JavaVersion.VERSION_1_8
}

repositories {
    mavenCentral()
}

def junitEngineVersion = '5.1.0'

dependencies {
    testCompile 'org.jmockit:jmockit:1.38'
    testImplementation "org.junit.jupiter:junit-jupiter-api:$junitEngineVersion"
    testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine:$junitEngineVersion"
}

test {
    useJUnitPlatform()
    testLogging {
        showStandardStreams = true
    }
}
```

#### Implementation classes:
```
public class Foo {
    private final Bar bar;

    public Foo(Bar bar) {
        this.bar = bar;
    }

    public String activate() {
        return "Active: [" + bar.bar() + "]";
    }
}
```

```
public class Bar {
    public String bar() {
        return "bar";
    }
}
```

#### Test class:
```
class FooTest {
    @Mocked
    private Bar bar;

    private Foo getFoo() {
        return new Foo(bar);
    }

    @Test
    void fooBar() {
        new Expectations() {
            {
                bar.bar();
                result = "baz";
                times = 1;
            }
        };
        Foo foo = getFoo();
        String activate = foo.activate();
        Assertions.assertEquals(activate, "Active: [baz]"); // Actually returns "Active: [null]"
    }
}
```

### Temporary workaround
As suggested by JMockit author.

```
test {
    useJUnitPlatform()
    def jmockit = configurations.testCompile.files.find { it.name.contains("jmockit") }.absolutePath
    jvmArgs "-javaagent:$jmockit"
}
```

**Comment by blindpirate** [comment]
Thanks for reporting. And thank @rliesenfeld a lot for the detailed analysis and solution.

**Comment by oehme** [comment]
I think this is also the cause for #4544

**Comment by oehme** [comment]
@rliesenfeld What do you mean by:

> The way this check is implemented has the bug, where classes get loaded unnecessarily (generally, doing so will mess up any Java agent which installs a ClassFileTransformer some time after JVM startup)

Why would loading an anonymous class mess up the Java agent? As far as I can see, it'll be loaded anyway as the test is executed.

**Comment by rliesenfeld** [comment]
@oehme I think the root problem here is that Gradle, for some reason, is not integrating properly with JUnit 5. Instead of searching for test classes itself, Gradle should delegate it to the [JUnit Platform Launcher API](https://junit.org/junit5/docs/current/user-guide/#launcher-api), which is meant to be used by IDEs and build tools such as Gradle. That way, the problem of loading potential test classes would be taken care of by JUnit itself.

**Comment by oehme** [comment]
We use the launcher API of course, but we still need to tell it which classes to analyse. I'm pretty sure if we pass it the name of an anonymous class (instead of the class object like we currently do) it would load it just like we do to find out whether it has potential tests. It's worth a try though.

**Comment by rliesenfeld** [comment]
From what I see in `JUnitPlatformTestClassProcessor`, it seems that Gradle is only using JUnit's `Launcher` for test execution (it calls `launcher.execute`), but not for test discovery (ie, it doesn't call `launcher.discover`, that I could find).

**Comment by blindpirate** [comment]
Just add some information here. Manually excluding classes which end with `$1` can solve this problem, but I'm reluctant to do this:

- `SomeClass$1` is a valid class name, there's no guarantee that it's an inner class
- I also think depending on class loading order is not a good idea. 

@rliesenfeld I'd like to provide some more context. The reason why Gradle loads and filters the test classes is #4363 #4325 #4492 . In short, currently we don't have a perfect solution for all these scenarios.

**Comment by oehme** [comment]
@rliesenfeld I think you misunderstand the JUnit launcher API. `launcher.execute` takes a discovery request. So it first discovers and then executes tests. `launcher.discover` only does discovery, but no execution, useful for when you only want to tell the user what is available.

@blindpirate Instead of instantiating the classes up-front, we can pass their names to the JUnit launcher and do our filtering using a TestFilter instead. This *might* fix the issue, so we could give it a try.

**Comment by rliesenfeld** [comment]
@oehme You're right, both `discover` and `execute` create a new `TestPlan` - I didn't expect that.
I still think that letting JUnit 5 do all the discovering, loading, and executing of test classes is the way to go. I wrote and executed the following code inside the JMockit project, running from IntelliJ with the necessary dependencies in main/pom.xml:

```
import org.junit.platform.engine.*;
import org.junit.platform.launcher.*;
import org.junit.platform.launcher.core.*;
import static org.junit.platform.engine.discovery.DiscoverySelectors.*;

public final class JUnit5TestRun
{
   public static void main(String[] args) {
      String testClassesDir = JUnit5TestRun.class.getProtectionDomain().getCodeSource().getLocation().getPath();

      LauncherDiscoveryRequest request = LauncherDiscoveryRequestBuilder.request().selectors(
         selectDirectory(testClassesDir), selectPackage("otherTests")
      ).build();
      Launcher launcher = LauncherFactory.create();

      launcher.execute(request, new TestExecutionListener() {
         @Override
         public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
            System.out.println(testIdentifier.getDisplayName() + ": " + testExecutionResult.getStatus());
         }
      });
   }
}
````
It works fine, no anonymous classes get loaded prematurely.

**Comment by sbrannen** [comment]
@rliesenfeld, 

`selectDirectory()` is not used by the JUnit Jupiter test engine.

If you want to restrict the classpath roots searched, you'll need to do something like the following.

```java
Path classpath = Paths.get(
    JUnit5TestRun.class.getProtectionDomain().getCodeSource().getLocation().toURI());

LauncherDiscoveryRequest request = LauncherDiscoveryRequestBuilder.request()
    .selectors(
        selectClasspathRoots(Collections.singleton(classpath)),
        selectPackage("otherTests")
    ).build();
```

**Comment by oehme** [comment]
The things is that we don't want JUnit to do classpath searching in the test worker. We assign test classes to run to multiple worker processes, so each process gets an explicit list of classes. This list of classes is built by the main daemon without loading them. Loading them could pollute the main Gradle daemon with misbehaving test classes (e.g. classes that create memory leaks). Thus we cannot use JUnit's scanning there, but instead use simple directory walking.

We could exclude anonymous classes from this basic directory walk. This would fix this issue and I can't see any adverse side-effects, since no test framework is gonna treat anonymous classes as an entry point.

**Comment by sbrannen** [comment]
> since no test framework is gonna treat anonymous classes as an entry point.

Yessssssssss.... let's hope so! 😉

**Comment by oehme** [comment]
I know, I should never say never ;)

**Comment by blindpirate** [comment]
@sbrannen @marcphilipp

Is it possible to filter class in test discovery phase, i.e. register a [ClassFilter](https://github.com/junit-team/junit5/blob/6b4de716ba3b0b46c847e340e2f68954d0cc6192/junit-platform-commons/src/main/java/org/junit/platform/commons/util/ClassFilter.java#L31)? Similar to [here](https://github.com/junit-team/junit5/blob/6b4de716ba3b0b46c847e340e2f68954d0cc6192/junit-platform-engine/src/main/java/org/junit/platform/engine/support/filter/ClasspathScanningSupport.java#L49). Currently seems we can only register class name filter via public API.

**Comment by marcphilipp** [comment]
> Is it possible to filter class in test discovery phase, i.e. register a ClassFilter?

Currently only `ClassNameFilter` and `PackageNameFilter` can be used. Even if you could pass a `Filter<Class<?>>`, JUnit would have to use reflection to load the class and pass it to the filter. Unless I'm missing something, that would defeat the purpose of not loading anonymous classes in the first place, wouldn't it?

I think excluding anonymous classes during Gradle's directory walk, as suggested by @oehme, should be safe and probably the way to go.

**Comment by blindpirate** [comment]
Thanks @marcphilipp . I was a bit reluctant to recognize an anonymous class via its name (`$` and ending number) but looks like it's kind of best solution so far.

**Comment by marcphilipp** [comment]
If I remember correctly, you use ASM to inspect classes. I would expect that it can tell you whether a class is anonymous, doesn't it?

**Comment by oehme** [comment]
We don't use ASM for JUnit 5. We only used it for JUnit 4 to scan for test classes, because if I remember correctly JUnit 4 would otherwise try to run things like abstract classes and fail. I'd like to avoid the ASM cost for JUnit 5, since it already does proper detection on its own.

I don't see a problem with looking for `$` followed by a number. We already have other code that assumes that `$` separates nested classes (e.g. our incremental compiler) and I have yet to see that a compiler violates that assumption.

**Comment by blindpirate** [comment]
Closed by #4774
